File: esserver/src/main/java/web/esserver/models/Scooter.java
package web.esserver.models;


import com.fasterxml.jackson.annotation.JsonView;


import javax.persistence.*;
import java.time.LocalDate;
import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.List;
import java.util.Random;

@Entity

@NamedQueries({
//  Find all methode
  @NamedQuery(name = "Scooter-find_all", query = "SELECT sc FROM Scooter sc"),

//  select waar status == aan de gegeven status
  @NamedQuery(name = "Scooter-find_by_status", query = "SELECT sc FROM Scooter sc WHERE sc.status =?1"),
  @NamedQuery(name = "Scooter-find_by_battery", query = "SELECT sc FROM Scooter sc WHERE sc.batteryCharger <=?1 "),
  @NamedQuery(name = "Trip_find_current_from_scooter", query = "SELECT sc FROM Scooter sc WHERE sc.status = 'INUSE' ")


})
public class Scooter {
  private static long lastId = 0;
  @JsonView(sumAll.class)
  @GeneratedValue()
  @Id
  private long id;
  @JsonView(sumAll.class)
  private String tag;

  @JsonView(sumAll.class)
  private String status;
  private String gpsLocation;

  @JsonView(sumAll.class)
  private int batteryCharger;
  private int mileage;

  @OneToMany(mappedBy = "scooter")
  private List<Trip> trips = new ArrayList<>();

  public List<Trip> getTrips() {
    return trips;
  }

  public void addTrips(Trip trip) {
    this.trips.add(trip);
    trip.setScooter(this);
  }

  public void removeTrips(Trip trip) {
    this.trips.remove(trip);

  }

  public static Trip startNewTrip(LocalDateTime startDateTime) {

    Trip trip = new Trip(0, "3", LocalDateTime.now(), LocalDateTime.now(), 33, 33);

    return trip;
  }

  public Scooter(long id, String tag, String status, String gpsLocation, int batteryCharge, int mileAge) {
    this.id = id;
    this.tag = tag;
    this.status = status;
    this.gpsLocation = gpsLocation;
    this.batteryCharger = batteryCharge;
    this.mileage = mileAge;
  }

  public Scooter() {

  }

  public static void setLastId(long lastId) {
    Scooter.lastId = lastId;
  }

  public void setId(long id) {
    this.id = id;
  }

  public void setTag(String tag) {
    this.tag = tag;
  }

  public void setStatus(String status) {
    this.status = status;
  }

  public void setGpsLocation(String gpsLocation) {
    this.gpsLocation = gpsLocation;
  }


  public int getBatteryCharger() {
    return batteryCharger;
  }

  public void setBatteryCharger(int batteryCharger) {
    this.batteryCharger = batteryCharger;
  }

  public void setMileage(int mileage) {
    this.mileage = mileage;
  }

  public int getMileage() {
    return mileage;
  }


  public static long getLastId() {
    return lastId;
  }

  public long getId() {
    return id;
  }

  public String getTag() {
    return tag;
  }

  public String getStatus() {
    return status;
  }

  public String getGpsLocation() {
    return gpsLocation;
  }


  public static Scooter createRandomScooter() {
    //id aanmaken

    long id = lastId++;

    //tag aanmaken
    Random r = new Random();
    String letters = "abcdefghijklmnopqrstuvwxyz";
    StringBuilder tagResult = new StringBuilder();
    final int N = 8;
    for (int i = 0; i < N; i++) {
      tagResult.append(letters.charAt(r.nextInt(letters.length())));
    }
    String tagResultFinal = tagResult.toString();

    //random status aanroepen
    double randomGetalEnum = Math.floor(Math.random() * 4) + 1;
    StringBuilder statusSb = new StringBuilder();

    if (randomGetalEnum < 2) {
      statusSb.append("INUSE");
    } else if (randomGetalEnum >= 2 && randomGetalEnum <= 3) {
      statusSb.append("IDLE");
    } else if (randomGetalEnum > 3) {
      statusSb.append("MAINTENANCE");
    }
    String statuss = statusSb.toString();

    //GPS locatie
    double longitudeGPS = 52379189;
    double eindeLong = Math.floor(Math.random() * 10);
    double longitude = longitudeGPS + eindeLong;

    double langitudeGPS = 4899431;
    double eindeLang = Math.floor(Math.random() * 10);
    double langitude = langitudeGPS + eindeLang;

    StringBuilder gpsLocatie = new StringBuilder();
    gpsLocatie.append(longitude + " " + langitude);
    String gps = gpsLocatie.toString();

    //batterij percentage pakken
    int batterijPercentage = (int) Math.floor(Math.random() * 100);
    int battery = batterijPercentage;
    if (batterijPercentage < 5) {
//   als het onder 5 is naar 5 maken;
      battery = 5;
    }
    // Aantal kilometers
    int distance = (int) Math.floor(Math.random() * 100);

    return new Scooter(lastId, tagResultFinal, statuss, gps, battery, distance);
  }


  public class sumAll {
  }


}

File: src/app/models/scooter.ts

export class Scooter {
  public status: scooterStatus;
  public gpsLocation: string;
  public id: number;
  public tag: string;
  public batteryCharger: number;
  public mileage: number;

  public scooter: Scooter[];
  public selected: Scooter;
  public displayScooter: Scooter;
  public selectedindex: number;

  static Copy(sc: Scooter): Scooter {
    return Object.assign(new Scooter(), sc);
  }

  public  equals(i): boolean {
    return this.id === i.id &&
      this.tag === i.tag &&
      this.gpsLocation === i.gpsLocation &&
      this.batteryCharger === i.batteryCharger &&
      this.status === i.status &&
      this.mileage === i.mileage;

  }



  public static createRandomScooter():Scooter{
    let sc = new Scooter();
    sc.id = 30001 + Math.floor(Math.random() * 100);
    sc.mileage = Math.floor(Math.random() * 1001);


    //Status
    const statusGetal = Math.floor(Math.random() * Math.floor(3) + 1);
    if (statusGetal === 1) {
      sc.status = scooterStatus.IDLE
    }
    if (statusGetal === 2) {
      sc.status = scooterStatus.INUSE
    }
    if (statusGetal === 3) {
      sc.status = scooterStatus.MAINTENANCE
    }
    //batterycharge
    var batteryPercent = Math.floor(Math.random() * 100);
    if (batteryPercent < 5) {
      sc.batteryCharger = 5;
    } else sc.batteryCharger = batteryPercent;

    var result = '';
    var characters = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789';
    var charactersLength = characters.length;
    for (var i = 0; i < 8; i++) {
      result += characters.charAt(Math.floor(Math.random() * charactersLength));
    }
    sc.tag = result;

    //gps
    var begin = 52379189
    var eind = Math.floor(Math.random() * 10);
    begin += eind
    var long = begin;

    var beginn = 4899431
    var eindn = Math.floor(Math.random() * 10);
    beginn += eindn

    var lat = beginn;

    sc.gpsLocation = lat + "N " + lat + "E"

    if (sc.status === scooterStatus.INUSE) {
      sc.gpsLocation = '';
    }

    return sc;
  }

}

export enum scooterStatus {
  IDLE = 'IDLE',
  INUSE = 'INUSE',
  MAINTENANCE = ' MAINTENANCE'
}










